### PR TITLE
[interop][SwiftToC++] print C++ interface for top-level Swift functions

### PR DESCRIPTION
**Pull request by hyp**
Emit C++ function thunks that correspond to Swift top-level functions, alongside their `extern "C"` function signatures, which are places in the `_impl` namespace in the module's namespace.

### CODE REVIEW

**Comment by hyp** [comment]
@swift-ci please test

File: lib/PrintAsClang/DeclAndTypePrinter.cpp [hunk 1]
```diff
@@ -2012,7 +2147,8 @@ class DeclAndTypePrinter::Implementation
 public:
   void print(Type ty, Optional<OptionalTypeKind> optionalKind,
              Identifier name = Identifier(),
-             IsFunctionParam_t isFuncParam = IsNotFunctionParam) {
+             IsFunctionParam_t isFuncParam = IsNotFunctionParam,
+             size_t *nameIndex = nullptr) {
```
Comment (zoecarver) [inline lib/PrintAsClang/DeclAndTypePrinter.cpp:2151]: This doesn't make sense as an API. Please use optional, or maybe a signed type. But optional is probably best.

File: lib/PrintAsClang/DeclAndTypePrinter.cpp [hunk 2]
```diff
@@ -739,22 +739,23 @@ class DeclAndTypePrinter::Implementation
...
+      size_t index = 1;
+      interleave(
+          *params,
+          [&](const ParamDecl *param) {
+            OptionalTypeKind kind;
+            Type objTy;
+            std::tie(objTy, kind) =
+                getObjectTypeAndOptionality(param, param->getInterfaceType());
+            print(objTy, kind, param->getName(), IsFunctionParam,
+                  printEmptyParamNames ? &index : nullptr);
```
Comment (zoecarver) [inline lib/PrintAsClang/DeclAndTypePrinter.cpp:755]: See below comment.

File: lib/PrintAsClang/DeclAndTypePrinter.cpp [hunk 3]
```diff
@@ -763,19 +764,145 @@ class DeclAndTypePrinter::Implementation
...
+    assert(!FD->getGenericSignature() &&
+           "top-level generic functions not supported here");
+    auto funcTy = FD->getInterfaceType()->castTo<FunctionType>();
+    auto resultTy =
+        getForeignResultType(FD, funcTy, asyncConvention, errorConvention);
+
+    os << "inline ";
+    printFunctionDeclAsCFunctionDecl(FD,
+                                     FD->getName().getBaseIdentifier().get(),
+                                     resultTy, /*printEmptyParamNames=*/true);
```
Comment (zoecarver) [inline lib/PrintAsClang/DeclAndTypePrinter.cpp:877]: Why only print empty param names some places? Seems like that should maybe be a global config. Then you also wouldn't have to launder an int down to `print`.

File: lib/PrintAsClang/CxxSynthesis.h [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+#ifndef SWIFT_PRINTASCLANG_CXXSYNTHESIS_H
+#define SWIFT_PRINTASCLANG_CXXSYNTHESIS_H
+
+#include "swift/Basic/LLVM.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/raw_ostream.h"
+
+namespace swift {
+
+namespace cxx_synthesis {
```
Comment (zoecarver) [inline lib/PrintAsClang/CxxSynthesis.h:22]: I don't think there's anything wrong with this, but usually, there'd be some object which we put all these functions in. Maybe `Implementation`?

File: lib/PrintAsClang/DeclAndTypePrinter.cpp [hunk 5]
```diff
@@ -763,19 +764,145 @@ class DeclAndTypePrinter::Implementation
...
+      // add a comment with a demangled function name.
+      os << " // ";
+      FD->getName().print(os);
+    }
+    os << "\n";
+    return funcABI;
+  }
+
+  void printAbstractFunctionAsCxxFunctionThunk(
+      FuncDecl *FD, const FuncionSwiftABIInformation &funcABI) {
```
Comment (zoecarver) [inline lib/PrintAsClang/DeclAndTypePrinter.cpp:861]:
```suggestion
      assert(outputLang == OutputLanguageMode::Cxx);
      FuncDecl *FD, const FuncionSwiftABIInformation &funcABI) {
```

File: lib/PrintAsClang/DeclAndTypePrinter.cpp [hunk 6]
```diff
@@ -763,19 +764,145 @@ class DeclAndTypePrinter::Implementation
...
+    bool useMangledSymbolName() const { return !isCDecl; }
+
+  private:
+    bool isCDecl;
+    std::string mangledName;
+    StringRef cdeclName;
+  };
+
+  FuncionSwiftABIInformation
+  printSwiftABIFunctionSignatureAsCxxFunction(FuncDecl *FD) {
```
Comment (zoecarver) [inline lib/PrintAsClang/DeclAndTypePrinter.cpp:825]: Maybe add a comment above this and `printAbstractFunctionAsCxxFunctionThunk` to explain roughly what they do and why you would use one vs the other.

File: lib/PrintAsClang/DeclAndTypePrinter.cpp [hunk 7]
```diff
@@ -763,19 +764,145 @@ class DeclAndTypePrinter::Implementation
...
+    printFunctionClangAttributes(FD, funcTy);
+    printAvailability(FD);
+    os << " {\n";
+    os << "  return " << cxx_synthesis::getCxxImplNamespaceName()
+       << "::" << funcABI.getSymbolName() << '(';
+
+    auto params = FD->getParameters();
+    if (params->size()) {
+      size_t index = 1;
+      interleave(
```
Comment (zoecarver) [inline lib/PrintAsClang/DeclAndTypePrinter.cpp:889]: Nit: `interleaveComma`

File: test/Interop/SwiftToCxx/functions/cdecl.swift [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+// RUN: %check-cxx-header-in-clang -std=c++14 %t/empty.h
+// RUN: %check-cxx-header-in-clang -std=c++17 %t/empty.h
+
+// CHECK-LABEL: namespace CdeclFunctions {
+
+// CHECK: namespace _impl {
+// CHECK: extern "C" int cfuncPassTwo(int x, int y) noexcept;
+// CHECK: }
+
+@_cdecl("cfuncPassTwo")
```
Comment (zoecarver) [inline test/Interop/SwiftToCxx/functions/cdecl.swift:14]: It's kind of interesting that this cdecl name is different from the declared fn, so maybe the name should be `differentCDeclName` or something like that. What do you think?

(Or maybe the test name?)

File: test/Interop/SwiftToCxx/functions/cdecl.swift [hunk 9]
```diff
@@ -0,0 +1,19 @@
+// RUN: %empty-directory(%t)
+// RUN: %target-swift-frontend(mock-sdk: %clang-importer-sdk) %s -typecheck -module-name CdeclFunctions -emit-cxx-header-path %t/empty.h
```
Comment (zoecarver) [inline test/Interop/SwiftToCxx/functions/cdecl.swift:2]: Do we need the mock sdk here?

File: test/Interop/SwiftToCxx/functions/cdecl.swift [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+// RUN: %FileCheck %s < %t/empty.h
+
+// RUN: %check-cxx-header-in-clang -std=c++14 %t/empty.h
+// RUN: %check-cxx-header-in-clang -std=c++17 %t/empty.h
```
Comment (zoecarver) [inline test/Interop/SwiftToCxx/functions/cdecl.swift:6]: We can't automate this to run 5-10 configs?

**Review by zoecarver — COMMENTED** [review]
Overall, looks really good. Excited to see the progress! I may have a couple more comments, but generally it looks good.

File: lib/PrintAsClang/DeclAndTypePrinter.cpp (hunk 1 above)
Comment (zoecarver) [reply lib/PrintAsClang/DeclAndTypePrinter.cpp:2151]: I was also thinking, maybe we could compute this in the caller and just pass it as `name`.

Comment (zoecarver) [reply lib/PrintAsClang/DeclAndTypePrinter.cpp:2151]: Another thought: this requires a bit of refactoring but `name` doesn't need to be an `Identifier` which might help. We could refactor it to be a `std::string` and in the caller just use `std::to_string`.

**Comment by zoecarver** [comment]
@LebJe, @hooray I noticed both of you have been putting emojis on the interop patches. If you're just watching from the side lines, that's totally fine (and thanks for the encouragement). But if you want to get involved in the project, please reach out to me (either on the forums or my email) and I can help you get started :)

**Comment by plotfi** [comment]
@zoecarver @hyp is this a better alternative to @_cdecl?

**Comment by zoecarver** [comment]
> @zoecarver @hyp is this a better alternative to @_cdecl?

I mean, "better" is really subjective. I think we may actually not take this path of "export everything," and instead require users to mark the functions they want exported. Maybe this can be one of the first working group discussions :)

File: lib/PrintAsClang/CxxSynthesis.h (hunk 4 above)
Comment (hyp) [reply lib/PrintAsClang/CxxSynthesis.h:22]: Sure, I ca pull it into an object class.

File: lib/PrintAsClang/DeclAndTypePrinter.cpp (hunk 3 above)
Comment (hyp) [reply lib/PrintAsClang/DeclAndTypePrinter.cpp:877]: we need to print out empty parameters for the inline thunks specifically as we need to use them in the body.

File: test/Interop/SwiftToCxx/functions/cdecl.swift (hunk 9 above)
Comment (hyp) [reply test/Interop/SwiftToCxx/functions/cdecl.swift:2]: Nope, removed.

File: test/Interop/SwiftToCxx/functions/cdecl.swift (hunk 10 above)
Comment (hyp) [reply test/Interop/SwiftToCxx/functions/cdecl.swift:6]: Good idea. I can do that as a follow-up.

File: test/Interop/SwiftToCxx/functions/cdecl.swift (hunk 8 above)
Comment (hyp) [reply test/Interop/SwiftToCxx/functions/cdecl.swift:14]: Yeah, that makes sense to me.

File: lib/PrintAsClang/DeclAndTypePrinter.cpp [hunk 11]
```diff
@@ -763,19 +764,145 @@ class DeclAndTypePrinter::Implementation
...
+    FuncionSwiftABIInformation(FuncDecl *FD, Mangle::ASTMangler &mangler) {
+      isCDecl = FD->getAttrs().hasAttribute<CDeclAttr>();
+      if (!isCDecl) {
+        mangledName = mangler.mangleAnyDecl(FD, /*prefix=*/true);
+      } else {
+        cdeclName = FD->getAttrs().getAttribute<CDeclAttr>()->Name;
+      }
+    }
+
+    StringRef getSymbolName() const {
```
[the remark this answers is not in the document]
Comment (hyp) [reply lib/PrintAsClang/DeclAndTypePrinter.cpp:810]: Good idea.

**Comment by hyp** [comment]
@zoecarver please take another look. I will need to make some changes in a follow-up PR related to unified header, but I would like to land this one first.

**Comment by hyp** [comment]
@swift-ci please test

**Comment by zoecarver** [comment]
So this is still exporting everything?

**Comment by hyp** [comment]
> everything

For now yes, I will need to resolve that in the future follow-up PRs

**Comment by hyp** [comment]
@swift-ci please test linux platform

**Comment by hyp** [comment]
@swift-ci please test windows platform

**Comment by hyp** [comment]
@swift-ci please test linux platform

**Comment by hyp** [comment]
@swift-ci please test

**Comment by hyp** [comment]
@swift-ci please test

**Comment by hyp** [comment]
@swift-ci please test linux platform

**Comment by hyp** [comment]
@swift-ci please test linux platform

**Comment by hyp** [comment]
@swift-ci please test linux platform

**Comment by hyp** [comment]
@swift-ci please smoke test

**Comment by hyp** [comment]
@swift-ci please smoke test

**Comment by hyp** [comment]
@swift-ci please smoke test linux platform

File: test/Interop/SwiftToCxx/functions/cdecl.swift (hunk 10 above)
Comment (hyp) [reply test/Interop/SwiftToCxx/functions/cdecl.swift:6]: Added support here: https://github.com/apple/swift/pull/41794

**Pull request merged**